src/jit/morph.cpp

```diff
@@ -12913,9 +12913,12 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 {
                     noway_assert(varTypeIsIntOrI(tree));
 
-                    tree->gtOp.gtOp2 = op2 = gtNewOperNode(GT_NEG, tree->gtType, op2); // The type of the new GT_NEG
-                                                                                       // node should be the same
-                    // as the type of the tree, i.e. tree->gtType.
+                    // The type of the new GT_NEG node should be the same
+                    // as the type of the tree, i.e. tree->gtType. Otherwise we may
+                    // sign-extend incorrectly in cases where the NEG node ends up
+                    // feeding directly into a cast, for example (byte)(0 - s_1)
+                    // where s_1 is a byte.
+                    tree->gtOp.gtOp2 = op2 = gtNewOperNode(GT_NEG, tree->gtType, op2); 
```

Comment:
I noticed that this was already fixed for a different case (`0 - s_1`). I'm not sure if there was a test for this already, but I have added it in my test case as well (and expanded on the comment)

Reply:
The use of `tree->gtType` is somewhat questionable. Since this basically transforms `op2` into `NEG(op2)` using the actual type of `op2` would be more appropriate - `gtNewOperNode(GT_NEG, genActualType(op2->TypeGet()), op2)`. 

Using `tree`'s type implicitly assumes that it is the same type as the tree operand(s) type. That assumption holds today for binary opers like `MUL` and `SUB` but perhaps that won't always be the case (e.g. it might be useful to get rid of a flag like `GTF_MUL_64RSLT` and instead explicitly represent a 32x32to64 bit multiplication in IR).

Reply:
Sure, I can make that change.

I'm curious: the ancestor tree (the parent of `tree`) would in that future case have its child change type. Would more changes not be necessary to adapt to this? In the case I am fixing here,  `tree` ends up being completely replaced by its `NEG` node (`op2`), so using the type of the child here means the type doesn't change.

Reply:
> I'm curious: the ancestor tree (the parent of tree) would in that future case have its child change type. Would more changes not be necessary to adapt to this? In the case I am fixing here, tree ends up being completely replaced by its NEG node (op2), so using the type of tree here means the type doesn't change.

Sorry, I don't understand. Can you provide some sort of clarifying example?

Reply:
In this case the tree is morphed in separate steps like this:
```
   ancestor
      |
     MUL <- `tree`
    /   \
   -1   ...

to

  ancestor
     |
    MUL  <- `tree`
    /   \
   1     NEG
          |
         ...

to

   ancestor
      |
     NEG
      |
     ...
```

Of course this only happens because the multiplication is by `-1`. So the `NEG` node created here replaces `tree` entirely. That (and the explicit comment) is why using the type of `tree` seemed right to me.

In the future, if a `MUL` node could take 64-bit operands, would this not need to further change anyway? Otherwise the ancestor's child changes type, which it perhaps wouldn't handle.

Reply:
> So the NEG node created here replaces tree entirely

That is the result of a separate transform. If `long MUL(int, int)` will ever be valid then any attempt to remove such a `MUL` node will require checking the relevant types and inserting widening casts as needed. But transforms that introduces `NEG` nodes won't ever be affected by such a change.

More generally - changing `op` to `unary(op)` with type = `actualType(op)` should always be valid since it preserves the original type. Using `op`'s parent type works under the assumption that the parent type is the same as the operand type. That assumption is true for `MUL` and many other binary opers but not for all - for example the type of a relop (EQ, NE etc.) oper can be different from the type of its operands.

Reply:
Right, makes sense. Thanks for the explanation.